core/vm/contract.go

```diff
@@ -151,3 +153,15 @@ func (self *Contract) SetCallCode(addr *common.Address, hash common.Hash, code [
 	self.CodeHash = hash
 	self.CodeAddr = addr
 }
+
+// SetVmError records concrete error that can cause the transaction or
+// top-level call to revert
+func (self *Contract) SetVmError(err error) {
+	if self.VmError != nil {
+		self.VmError = err
+	}
+	// affect parent recursively
+	if parent, ok := self.caller.(*Contract); ok {
+		parent.SetVmError(err)
+	}
+}
```

Comment:
I don't think you need this or the contract.VMError field. All the Call* and Create methods already return the same error, and we only care about the outermost one, so internal ones can be discarded without any problems.